Detect current network and netmask

baelish opened the pull request:
@jfelten Please consider this pull request.

#### What this PR does / why we need it:
This PR detects the IP address and Netmask correctly. Which results in only valid routes being passed to the clients.

Previously the network was guessed (assumed it was a /24 by replacing the last octet with a 0) and then later the netmask was hardcoded to 255.255.240.0 as a result the requested route was almost always incorrect. As most people supply their own or use the vpn tunnel as a default gateway it was probably rare to cause any issues.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #13211

#### Special notes for your reviewer:
the `cidr2mask()` function was copied from [here](https://forum.archive.openwrt.org/viewtopic.php?id=47986&p=1#p220781) the rest was written from scratch.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

baelish committed:
Detect current network and netmask

Previously the network was guessed (assumed it was a /24 by replacing the last octet with a 0) and then later the netmask was hardcoded to 255.255.240.0 as a result the requested route is almost always incorrect. As most people supply their own or use the vpn tunnel as a default gateway it's probably rare to cause any issues.

Signed-off-by: Harold Drost <baelish@bluecell.net>

jasongwartz commented on stable/openvpn/templates/config-openvpn.yaml:
```diff
@@ -65,9 +94,16 @@ data:
+      int="${intAndIP%-*}"
+      ip="${intAndIP#*-}"
+
+      echo $int "and" $ip
```
intentional or leftover debug?

baelish replied:
130fbbb leftover debug

jasongwartz commented on stable/openvpn/templates/config-openvpn.yaml:
```diff
@@ -78,6 +114,7 @@ data:
       sed 's|OVPN_K8S_SEARCH|'"${FORMATTED_SEARCH}"'|' -i /etc/openvpn/openvpn.conf
       sed 's|OVPN_K8S_DNS|'"${DNS}"'|' -i /etc/openvpn/openvpn.conf
       sed 's|NETWORK|'"${NETWORK}"'|' -i /etc/openvpn/openvpn.conf
+      sed 's|NETMASK|'"${NETMASK}"'|' -i /etc/openvpn/openvpn.conf
```
maybe I'm missing something here, but I'm guessing we want to put the NETMASK in the openvpn.conf file somewhere? I don't see `NETMASK` in the openvpn.conf file as a placeholder at the moment - are you using `.Values.openvpn.OVPN_SUBNET` in its place?

baelish replied:
db363d6 Well spotted, embarrasingly my testing didn't catch that as the helm upgrade I performed didn't rewrite the openvpn.conf file which I had manually edited to contain the correction in the mentioned commit.

jasongwartz commented on stable/openvpn/templates/config-openvpn.yaml:
```diff
@@ -53,6 +53,35 @@ data:
+              if [ $i -le $(( mask / 8)) ]; then
+                  netOctets="$netOctets.$octet"
+              elif [ $i -eq  $(( mask / 8 +1 )) ]; then
+                  echo $octet
```
when I run `cidr2net()` locally, my output is:
```
63
10.0.3.0
```
which makes sense, given that my IP is 10.0.3.63. However, I suspect this `echo $octet` might be leftover debug?

baelish replied:
130fbbb Leftover debug.

jasongwartz reviewed (commented):
Thanks for the contribution!

This isn't a situation I personally encounter with this chart, so I have a few general questions but in general am just comparing what you've implemented here with the openvpn docs.
Cheers!

baelish committed:
Removing unnecessary debug messages.

Signed-off-by: Harold Drost <baelish@bluecell.net>

baelish committed:
Actually fixing the thing it was all about... :$

My testing worked because helm believed the `openvpn.conf` unchanged and therefore didn't replace the manually edited file which had the change in this commit.
Tested using brand new deployment now works correctly and without routing errors.

Signed-off-by: Harold Drost <baelish@bluecell.net>
